### PR TITLE
feat: configurable image hosts

### PR DESCRIPTION
**Pull request by dordsor21**

### CODE REVIEW

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/util/image/ImageUtil.java [hunk 1]
```diff
@@ -229,7 +230,9 @@ public static URI getImageURI(String arg) throws InputParseException {
                 if (arg.contains("imgur.com") && !arg.contains("i.imgur.com")) {
                     arg = "https://i.imgur.com/" + arg.split("imgur.com/")[1] + ".png";
                 }
-                return new URL(arg).toURI();
+                URL url = new URL(arg);
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/fastasyncworldedit/core/util/image/ImageUtil.java:233]: This constructor is deprecated in Java 20, maybe it makes sense to direclty construct the URI and also let the checkImageHost take a URI?

**Comment by PierreSchwang** [comment]
Maybe we should directly support glob patterns and not limit on hosts directly?

**Comment by dordsor21** [comment]
> Maybe we should directly support glob patterns and not limit on hosts directly?

Could do, but not sure if that's more effort than it's worth tbh. Could count on one hand the number of people specifically requesting this haha

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/util/image/ImageUtil.java (hunk 1 above)
Comment (dordsor21) [reply worldedit-core/src/main/java/com/fastasyncworldedit/core/util/image/ImageUtil.java:233]: OIC, yeah probably worth that then

**Review by NotMyFault — COMMENTED** [review]
I don't think adding, even if configurable, settings like these are a great idea. Configured improperly, there are many possible drawbacks.

**Comment by dordsor21** [comment]
It's more with the mind of build team/whitelist servers where it might be quite beneficial to be able to use images from anywhere, rather than having to reupload to imgur every time

**Comment by OneLiteFeather** [comment]
You could also self host an upload solution like an S3 Storage or "picture" server so you could use your own link for that, right?

**Comment by NotMyFault** [comment]
> It's more with the mind of build team/whitelist servers

In such a case, the implementation should be API only, allowing the people needed to extend it on their own.

**Comment by dordsor21** [comment]
Tbh that's too much work for the user, especially given most users don't touch their configs at all in the first place. The warning's there, and FAWE has never shied from assuming a reasonable level of intelligence in its users (and I don't want to start making FAWE for dummies yet tbh). And besides, I'd rather a server OOM be someone's kick to implement reasonable security procedures than an actually dangerous exploit

File: worldedit-core/src/main/java/com/fastasyncworldedit/core/configuration/Settings.java [hunk 2]
```diff
@@ -671,6 +671,14 @@ public static class WEB {
         })
         public int MAX_IMAGE_SIZE = 8294400;
 
+        @Comment({
+                "Whitelist of hostnames to allow images to be downloaded from",
+                " - Adding '*' to the list will allow any host, but this is NOT adviseable",
+                " - Crash exploits exist with malphormed images",
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/fastasyncworldedit/core/configuration/Settings.java:677]:
```suggestion
                " - Crash exploits exist with malformed images",
```

**Review by SirYwell — COMMENTED** [review]
Small typo, but looks good otherwise. This will also help people who do not trust imgur and rather trust their own services.

**Review by SirYwell — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
Can still revert it if it causes major issues 🤷

**Pull request merged**